protocol-units/da/m1/light-node-verifier/src/celestia/mod.rs

```diff
@@ -1,79 +1,70 @@
-use crate::Verifier;
-use celestia_rpc::{BlobClient, Client, HeaderClient};
+use crate::{Error, Verified, VerifierOperations};
+use celestia_rpc::Client;
+#[cfg(feature = "pessimistic")]
+use celestia_rpc::{BlobClient, HeaderClient};
 use celestia_types::{nmt::Namespace, Blob};
-use m1_da_light_node_grpc::VerificationMode;
+use m1_da_light_node_util::inner_blob::InnerBlob;
 use std::sync::Arc;
 
 #[derive(Clone)]
-pub struct V1Verifier {
+pub struct Verifier {
 	pub client: Arc<Client>,
 	pub namespace: Namespace,
 }
 
-#[tonic::async_trait]
-impl Verifier for V1Verifier {
-	/// All verification is the same for now
-	async fn verify(
-		&self,
-		_verification_mode: VerificationMode,
-		blob: &[u8],
-		height: u64,
-	) -> Result<bool, anyhow::Error> {
-		let celestia_blob = Blob::new(self.namespace.clone(), blob.to_vec())?;
-
-		celestia_blob.validate()?;
-
-		// wait for the header to be at the correct height
-		self.client.header_wait_for_height(height).await?;
-
-		// get the root
-		let dah = self.client.header_get_by_height(height).await?.dah;
-		let root_hash = dah.row_root(0).ok_or(anyhow::anyhow!("No root hash found"))?;
-
-		// get the proof
-		let proofs = self
-			.client
-			.blob_get_proof(height, self.namespace.clone(), celestia_blob.commitment)
-			.await?;
-
-		// get the leaves
-		let leaves = celestia_blob.to_shares()?;
-
-		// check if included
-		for proof in proofs.iter() {
-			proof
-				.verify_complete_namespace(&root_hash, &leaves, self.namespace.into())
-				.map_err(|e| anyhow::anyhow!("Failed to verify proof: {:?}", e))?;
-		}
-
-		Ok(true)
+impl Verifier {
+	pub fn new(client: Arc<Client>, namespace: Namespace) -> Self {
+		Self { client, namespace }
 	}
+}
 
-	async fn verify_cowboy(
-		&self,
-		_verification_mode: VerificationMode,
-		_blob: &[u8],
-		_height: u64,
-	) -> Result<bool, anyhow::Error> {
-		unimplemented!()
-	}
+#[tonic::async_trait]
+impl VerifierOperations<Blob, InnerBlob> for Verifier {
+	/// Verifies a Celestia Blob as a Valid InnerBlob
+	async fn verify(&self, blob: Blob, _height: u64) -> Result<Verified<InnerBlob>, Error> {
+		//@l-monninger: the light node itself does most of the work of verify blobs. The verification under the feature flag below is useful in zero-trust environments.
+		#[cfg(feature = "pessimistic")]
```

Comment:
Changing behavior with features is an antipattern: features are additive in a workspace, so if this library has a client A that expects the `not(feature = "pessimistic")` behavior, but there is another crate B that enables the "pessimistic" feature, the client A may get unexpected results or unwanted performance hit.
I think it's better to have a separate method that invokes the pessimistic verification.

Reply:
I thought you were going to say that. I'll split it out into another verifier.